jupyter_server/services/kernels/kernelmanager.py

```diff
@@ -711,4 +715,103 @@ class ServerKernelManager(AsyncIOLoopKernelManager):
         None, allow_none=True, help="The current execution state of the kernel"
     )
     reason = Unicode("", help="The reason for the last failure against the kernel")
+
     last_activity = Instance(datetime, help="The last activity on the kernel")
+
+    # A list of pathlib objects, each pointing at an event
+    # schema to register with this kernel manager's eventlogger.
+    # This trait should not be overridden.
+    @property
+    def core_event_schema_paths(self) -> t.List[pathlib.Path]:
+        return [DEFAULT_EVENTS_SCHEMA_PATH / "kernel_actions" / "v1.yaml"]
+
+    # This trait is intended for subclasses to override and define
+    # custom event schemas.
+    extra_event_schema_paths = List(
+        default_value=[],
+        help="""
+        A list of pathlib.Path objects pointing at to register with
+        the kernel manager's eventlogger.
+        """,
+    ).tag(config=True)
+
+    event_logger = Instance(EventLogger)
+
+    @default("event_logger")
+    def _default_event_logger(self):
+        """Initialize the logger and ensure all required events are present."""
+        if self.parent is not None and hasattr(self.parent, "event_logger"):
+            logger = self.parent.event_logger
+        else:
+            # If parent does not have an event logger, create one.
+            logger = EventLogger()
+        # Ensure that all the expected schemas are registered. If not, register them.
+        schemas = self.core_event_schema_paths + self.extra_event_schema_paths
+        for schema_path in schemas:
+            # Try registering the event.
+            try:
+                logger.register_event_schema(schema_path)
+            # Pass if it already exists.
+            except SchemaRegistryException:
+                pass
+        return logger
+
+    def emit(self, schema_id, data):
+        """Emit an event from the kernel manager."""
+        self.event_logger.emit(schema_id=schema_id, data=data)
+
+    @wraps(AsyncIOLoopKernelManager.start_kernel)
+    async def start_kernel(self, *args, **kwargs):
+        result = await super().start_kernel(*args, **kwargs)
+        self.emit(
+            schema_id="https://events.jupyter.org/jupyter_server/kernel_actions/v1",
+            data={
+                "kernel_id": self.kernel_id,
+                "kernel_name": self.kernel_name,
+                "action": "start",
+                "msg": f"Kernel {self.kernel_id} was started.",
+            },
+        )
+        return result
+
+    @wraps(AsyncIOLoopKernelManager.interrupt_kernel)
```

Comment:
I don't understand what this decorator buys us (but I find decorators confusing anyway).  Could you please explain its purpose?

Reply:
This ensures that the method "inherits" the same signature and docstring as the parent/overriden method. So the same arguments and keyword arguments are enforced. If the parent method's signature changes, this method automatically captures those changes as well. Further, the parent's docstring is inherited, which is handy as well.

Reply:
I see - thank you, that's helpful.  So, in theory, this _should_ be done for all subclasses?  I use `@overrides` for this in [gateway provisioners](https://github.com/jupyter-server/gateway_provisioners/blob/ec9b6952af006492d587e8e0ac0f0de09a44a805/gateway_provisioners/container.py#L120) which seems a bit cleaner and less confusing (IMHO).

Reply:
Ooo I like that. It does add a (temporary) dependency, but it looks like there is an accepted PEP adding this to the standard lib: https://peps.python.org/pep-0698/

I'll go ahead an update.